### Joint SIB and NBIS/SciLifeLab Autumn School

# Single Cell Analysis Course

<br/>

### October 2019


<br/>

### <img border="0" src="https://www.svgrepo.com/show/20800/event-date-and-time-symbol.svg" width="40" height="40"> Date

***

13.10.2019 - 18.10.2019

<br/>

### <img border="0" src="https://www.svgrepo.com/show/4199/placeholder-on-a-map.svg" width="40" height="40"> Location

***

Hotel Central Residence & Spa, Leysin, Switzerland.

Arrival: Sunday before 18h

Departure: Friday around 14h30

<br/>

## <img border="0" src="https://www.svgrepo.com/show/158264/schedule.svg" width="40" height="40"> Programme

***

For the full schedule, please go to: **[Schedule](schedule.md)**

#### Sunday

* ≈17-18h: arrival of the participants, check-in
* Welcome and dinner

#### Monday: Transcriptomics

* Introductory lecture (Vincent Gardeux)
* Quantification, QC & Normalization (Davide Risso)
* Dimensionality reduction (Paulo Czarnewski)

#### Tuesday: Transcriptomics

* Batch correction (Panagiotis Papasaikas)
* Clustering - methods overview (Charlotte Soneson)
* Cell fate mapping and trajectories (Wouter Saelens)

#### Wednesday: Transcriptomics

* Differential expression (Charlotte Soneson)
* Afternoon: Team activity
* Evening: Keynote lecture: Alejandro Sifrim, Laboratory of Reproductive Genomics, KU Leuwen, Belgium.

#### Thursday: Proteomics

* Transcriptome + proteome (Johan Reimegård)
* Differential abundance and differential state analysis of single cell cytometry data (Mark Robinson & Helena Crowell)

#### Friday: Other omics and integration

* Data integration methods (Sebastien Smallwood)
* Spatial mapping of scRNAseq (Lars Borm)
* Lineage tracing and scRNA (Maria Florescu)

<br/>

## <img border="0" src="https://www.svgrepo.com/show/410/list.svg" width="40" height="40"> Learning objectives

***

After this course you will be able to:
- use a range of bioinformatics tools to analyze single cell RNA-seq data
- discuss a variety of aspects of single cell RNA-seq data analysis
- understand the advantages and limitations of single cell RNA-seq data analysis

<br/>

## <img border="0" src="https://www.svgrepo.com/show/38706/group-of-people.svg" width="40" height="40"> Lecturers

***

- Alejandro Sifrim, Laboratory of Reproductive Genomics, KU Leuwen, Belgium
- Åsa Björklund, NBIS, Uppsala University, SciLifeLab, Uppsala, Sweden
- Charlotte Soneson, Computational Biology Group, Friedrich Miescher Institute for Biomedical Research / SIB, Basel, Switzerland
- Davide Risso, Department of Statistical Sciences, University of Padova, Italy
- Helena Crowell, Statistical Genomics, University of Zurich / SIB, Zurich, Switzerland
- Johan Reimegård, NBIS, Uppsala University, SciLifeLab, Uppsala, Sweden
- Lars Borm, Department of Medical Biochemistry and Biophysics, Karolinska Institutet, Stockholm, Sweden
- Maria Florescu, Hubrecht Institute, Developmental Biology and Stem Cell Research, Utrecht, Netherlands
- Mark Robinson, Statistical Genomics, University of Zurich / SIB, Zurich, Switzerland
- Nadine Zangger, Bioinformatics Core Facility, SIB, Lausanne, Switzerland
- Panagiotis Papasaikas, Computational Biology Group, Friedrich Miescher Institute for Biomedical Research / SIB, Basel, Switzerland
- Paulo Czarnewski, NBIS, Stockholm University, SciLifeLab, Stockholm, Sweden
- Sebastien Smallwood, Friedrich Miescher Institute for Biomedical Research, Basel, Switzerland
- Vincent Gardeux, Laboratory of Systems and Genetics, EPFL / SIB, Lausanne, Switzerland
- Wouter Saelens, VIB-UGent Center for Inflammation Research, Gent, Belgium

### Additional information

Coordination: Björn Nystedt (NBIS/SciLifeLab), Grégoire Rossier (SIB).

Scientific Committe: Åsa Björklund (NBIS/SciLifeLab), Michael Stadler & Charlotte Soneson (SIB and FMI Basel), Vincent Gardeux (EPFL).

For more information, please contact Grégoire Rossier.
